### PR TITLE
chore: prevent multiple CI runs when using pushed branched

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

Currently, when pushing a branch and opening a PR, you get multiple runs.
This makes it so only one happens.
https://stackoverflow.com/questions/58895283/stop-already-running-workflow-job-in-github-actions

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by fubuloubu** [comment]
This seems to work great! closing #1004

**Review by fubuloubu — APPROVED** [review]

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -2,6 +2,10 @@ on: ["push", "pull_request"]
 
 name: Test
 
+concurrency:
+  group: ${{ github.head_ref }}
```
Comment (fubuloubu) [inline .github/workflows/test.yaml:6]: someone gave the suggestion to use `github.ref` instead

**Comment by fubuloubu** [comment]
Okay I guess `github.ref` doesn't work

**Comment by unparalleled-js** [comment]
> Okay I guess `github.ref` doesn't work

Reverted!

Another nice thing about this that I noticed is that if you happen to push again before the workflow finishes, it will cancel it and only focus on the new one.

**Comment by unparalleled-js** [comment]
So it seems like when merging to main, the tests don't run again. But is that actually a problem? Because we require all branches to be caught up with main before they can merge anyway, so maybe its seen as a duplicate.

**Comment by unparalleled-js** [comment]
Ah... Found the problem:

```
Invalid workflow file
The workflow is not valid. .github/workflows/test.yaml (Line: 7, Col: 10): Unexpected type of value '', expected type: String.
```

**Comment by fubuloubu** [comment]
> Ah... Found the problem:
> 
> ```
> Invalid workflow file
> The workflow is not valid. .github/workflows/test.yaml (Line: 7, Col: 10): Unexpected type of value '', expected type: String.
> ```
> 
> The `.ref` thing does not exist on that event perhaps.

Ah yes, so maybe it needs to detect if it is a pull request or it is a branch push

File: .github/workflows/test.yaml (hunk 1 above)
Comment (fubuloubu) [inline .github/workflows/test.yaml:6]: Vyper is using `${{ github.workflow }}-{{ github.head_ref }}` here:

https://github.com/vyperlang/vyper/pull/3179

**Comment by unparalleled-js** [comment]
It seems like we are getting duplicate runs again, after copying Vyper's approach.

**Review by NotPeopling2day — APPROVED** [review]

**Pull request merged**